lib/media/media_source_engine.js

```diff
@@ -2052,6 +2052,22 @@ shaka.media.MediaSourceEngine = class {
 
     // Current/new codecs base and basic type match then no need to switch
     if (currentCodec === newCodec && currentBasicType === newBasicType) {
+      if (this.transmuxers_[contentType] && !transmuxer) {
+        this.transmuxers_[contentType].destroy();
+        delete this.transmuxers_[contentType];
+      } else if (!this.transmuxers_[contentType] && transmuxer) {
+        this.transmuxers_[contentType] = transmuxer;
+      } else if (transmuxer) {
+        // Compare if the transmuxer is different
+        if (this.transmuxers_[contentType] &&
```

Comment:
Whenever `codecSwitchIfNecessary_` detects that a transmuxer is needed, it creates a new instance of the desired transmuxer. It never recycles the old transmuxer. So doesn't that mean that this is always going to delete the old transmuxer?

I was going to say that we should be comparing the respective `shaka.extern.TransmuxerPlugin` to see if the two transmuxers are of the same type, instead of seeing if they are the same object. However, looking at the transmuxers we have now, none of them look to be stateless. Most of them have internal variables like `frameIndex_` which maybe we do want to reset by remaking the transmuxer.

If that is the case, and we never want to keep the old transmuxer, this whole segment could be greatly simplified.

```js
if (this.transmuxers_[contentType]) {
  this.transmuxers_[contentType].destroy();
  delete this.transmuxers_[contentType];
}
if (transmuxer) {
  this.transmuxers_[contentType] = transmuxer;
}
```

Reply:
yes we want to keep the transmuxer, because then we don't have to generate the initialization segment again

Reply:
Oh, also I see now that this isn't comparing the transmuxers themselves, it's comparing their `transmux` methods. I guess that works as a way to ensure they came from the same plugin.